File: Python/1026.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

# Save path values in a list
class Solution:
    def dfs(self, root, path):
        if not root: # If root is null
            # Update the returnVal of the node
            self.returnVal = max(self.returnVal, max(path) - min(path)) 
            return
        # If root is not null
        self.dfs(root.left, path + [root.val]) # O(lgn)
        self.dfs(root.right, path + [root.val]) # O(lgn)
        
    def maxAncestorDiff(self, root: TreeNode) -> int:
        # use Depth Tree First Search algorithm
        # Want abs(diff) -> least value is 0
        self.returnVal = 0
        self.dfs(root, [])
        return self.returnVal

# One value in a list -> list is necessary, otherwise assignment issue happens
class Solution:
    def maxAncestorDiff(self, root: TreeNode) -> int:
        def dfs(node: TreeNode, max_: int, min_: int) -> None:
            if not node: # If root is null
                return
            # If root is not null
            max_ = max(max_, node.val)
            min_ = min(min_, node.val)
            ret[0]  = max(ret[0], max_- min_)

            dfs(node.left, max_, min_) # O(lgn)
            dfs(node.right, max_, min_) # O(lgn)

        # use Depth Tree First Search algorithm
        # Want abs(diff) -> least value is 0
        ret = [0]
        dfs(root, -float("inf"), float("inf"))
        return ret[0]
        

File: CPP/1026.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    void diff(TreeNode* root, int max_, int min_, int *ret){
            if (root == NULL) return ;
            max_ = max(root->val, max_);
            min_ = min(root->val, min_);
            ret[0] = max(ret[0], max_ - min_);
            
            diff(root->left, max_, min_, ret);
            diff(root->right, max_, min_, ret);
    }
    
    int maxAncestorDiff(TreeNode* root) {
        int ret[1] = {0}, max_ = INT_MIN, min_ = INT_MAX;
        diff(root, max_, min_, ret);
        return ret[0];
    }
};